### ISSUE TITLE
Indecipherable error when starting a message queue handler - broken redis error handling

### ISSUE THREAD

**Issue by j0ni** [ptaoussanis/carmine#87]
Hi

I'm perplexed by this and would appreciate your help deciphering the error below.

```
2014-Jun-03 17:39:08 +0200 oden INFO [oden.redis] - Starting redis queue worker: {:pool {}, :spec {:host "localhost", :port 6379}}
2014-Jun-03 17:39:08 +0200 oden INFO [taoensso.carmine.message-queue] - Message queue worker starting: oden
2014-Jun-03 17:26:23 +0200 oden ERROR [taoensso.carmine.message-queue] - Worker error! Will backoff & retry.
java.lang.UnsupportedOperationException: nth not supported on this type: ExceptionInfo
                                     clojure.lang.RT.nthFrom                  RT.java:  857
                                         clojure.lang.RT.nth                  RT.java:  807
                      taoensso.carmine.message-queue/handle1        message_queue.clj:  277
taoensso.carmine.message-queue.Worker/start-polling-loop!/fn        message_queue.clj:  358
   taoensso.carmine.message-queue.Worker/start-polling-loop!        message_queue.clj:  342
                 taoensso.carmine.message-queue.Worker/fn/fn        message_queue.clj:  370
                         clojure.core/binding-conveyor-fn/fn                 core.clj: 1910
                                       clojure.lang.AFn.call                 AFn.java:   18
                         java.util.concurrent.FutureTask.run          FutureTask.java:  266
           java.util.concurrent.ThreadPoolExecutor.runWorker  ThreadPoolExecutor.java: 1142
          java.util.concurrent.ThreadPoolExecutor$Worker.run  ThreadPoolExecutor.java:  617
                                        java.lang.Thread.run              Thread.java:  745
```

I've wrapped the handler setup and the handler function itself in try/catch code, like this:

```clojure
(defn make-worker [{:keys [server-config thread-count qname worker]} handler]
  (try
    (let [handler (fn [{:keys [message]}]
                    (try
                      (let [message (edn/read-string message)
                            {:keys [metric timestamp rule-key]} message
                            result (handler metric timestamp rule-key)]
                        (if result {:status :success} {:status :error}))
                      (catch Throwable t (timbre/log :error (str "Handler error: " (.getMessage t))))))]
      (timbre/log :info (str "Starting redis queue worker: " server-config))
      (reset! worker (car-mq/worker server-config qname {:handler handler :nthreads thread-count})))
    (catch Throwable t (timbre/log :error (str "Failed to start handler: " (.getMessage t))))))
```

but I only see the error above. This all works in my local dev environment - this is in an ubuntu 12.04.2 VM. Same JDK version (1.8) and the same uberjar. I've verified that I can connect using redis-cli to localhost without any credentials, on the default port.

Any ideas?

**Comment by gsnewmark** [comment]
@ptaoussanis Hi! So, is exception described in the initial post the intended behavior of the queue worker when Redis returns error string as a result of the Lua script? 

Right now `poll-reply` [here](https://github.com/ptaoussanis/carmine/blob/master/src/taoensso/carmine/message_queue.clj#L244) would be an `ExceptionInfo` in case Redis returns error for the command, because `wcar` doesn't throw errors when multiple commands are passed ([related test](https://github.com/ptaoussanis/carmine/blob/d1c0df7e8acb41a240465eff3393f19202b4221f/test/taoensso/carmine/tests/main.clj#L84), [multiple command result parsing](https://github.com/ptaoussanis/carmine/blob/5d33e8444595756c54ea6b48d335535d7e95c287/src/taoensso/carmine/protocol.clj#L271)). For example, if Redis exhausts all allocated memory `poll-reply` would be an `ExceptionInfo` with following cause `ERR Error running script (call to f_c09905edc7fdcc40f471cef5920114821719103b): -OOM command not allowed when used memory > 'maxmemory'.   `, but [`handle1`](https://github.com/ptaoussanis/carmine/blob/d1c0df7e8acb41a240465eff3393f19202b4221f/src/taoensso/carmine/message_queue.clj#L175) would try to destructure it as vector leading to the initially described exception.

In any case, worker can't really work if memory is exhausted, but thrown exception could probably be a bit more helpful. Moreover, there're probably more cases when Redis returns error for the worker's Lua script, and showing actual Redis exception could really help in debugging the problem.

**Comment by ptaoussanis** [comment]
@gsnewmark Hi Ivan, thanks for the report. Will try take a look at this w/in the next week.

**Comment by ptaoussanis** [comment]
@gsnewmark Will be fixed with v2.13.0 - appreciate the detailed report, was helpful. Cheers :-)

**Comment by gsnewmark** [comment]
@ptaoussanis great, thanks!
